File: 0695-max-area-of-island/solution.cpp
class Solution {
public:
  int dfs(vector<vector<int>>& g, int i, int j) {
    if (i >= 0 && j >= 0 && i < g.size() && j < g[i].size() && g[i][j] == 1) {
      g[i][j] = 2;
      return 
        dfs(g, i-1, j) +
        dfs(g, i+1, j) +
        dfs(g, i, j-1) +
        dfs(g, i, j+1) + 1;
    }
    return 0;
  }

  int maxAreaOfIsland(vector<vector<int>>& g) {
    int m = 0;
    for(int i = 0; i < g.size(); i++)
      for(int j = 0; j < g[i].size(); j++)
        if (g[i][j] == 1) m = max(m, dfs(g, i, j));
    return m;
  }
};

File: 0695-max-area-of-island/solution.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROWS, COLS, result = len(grid), len(grid[0]), 0

        def dfs(r: int, c: int) -> int:
            if (
                r not in range(0, ROWS) or
                c not in range(0, COLS) or
                grid[r][c] != 1
            ):
                return 0

            grid[r][c] = 2

            return 1 + dfs(r - 1, c) + dfs(r + 1, c) + dfs(r, c - 1) + dfs(r, c + 1)

        for r in range(ROWS):
            for c in range(COLS):
                result = max(result, dfs(r, c))

        return result

